### PR TITLE
fmt: impl fmt::Pointer for smart pointer types

### LINKED ISSUE #24091: Smart pointer types should have a fmt::Pointer implementation
Types like Arc, Rc, and other smart pointers should have an implementation of `fmt::Pointer` so that the `{:p}` format specifier works

### PR DESCRIPTION
**Pull request by richo**
~~I believe this should fix the issue. Opening a PR to ensure noone duplicates effort, I'm running check now.~~

Closes #24091

### CODE REVIEW

**Comment by richo** [comment]
This builds and does something reasonable on my machine. Is it worth adding a test?

Also, are there any other smart pointer types that I should impl for? The only other one that looks relevant is `P` in rustc that would be useful for shotgun debugging, and perhaps `Box` ?

**Comment by richo** [comment]
Bonus impls for the other smart pointers I could find

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -276,6 +276,13 @@ impl<T: fmt::Debug + ?Sized> fmt::Debug for Box<T> {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
+impl<T> fmt::Pointer for Box<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Pointer::fmt(&*self, f)
```
Comment (alexcrichton) [inline src/liballoc/boxed.rs]: I think that this is infinitely recursive, did you mean `&**self`?

**Comment by alexcrichton** [comment]
Could you add a smoke test for the `Box`, `Rc`, and `Arc` impls as well?

File: src/liballoc/boxed.rs (hunk 1 above)
Comment (richo) [inline src/liballoc/boxed.rs]: Oops. Is `Box` specialcased in some way in the compiler? I can't seem to extract the `Unique` within using tuple access or a destructuring match.

```
expected `&Box<T>`,
    found `boxed::Box<_>`
(expected &-ptr,
    found struct `boxed::Box`) [E0308]
/Users/richo/code/ext/rust/src/liballoc/boxed.rs:282             Box(uniq) => fmt::Pointer::fmt(&uniq, f)
```

**Comment by richo** [comment]
Sure, I'll work out how I goofed box and add the test.

File: src/liballoc/boxed.rs (hunk 1 above)
Comment (alexcrichton) [inline src/liballoc/boxed.rs]: Yeah `Box` is pretty special, and I'm not actually sure the definition has much bearing on what you can do with the type itself. You'll probably want to pull out `&T` from `&Box<T>` and then format that.

**Comment by richo** [comment]
I added a test, and made it pad pointers to native pointer width if no format is specified, which I bellieve to be correct. Let me know if I should rip that off though.

Make check is running on my machine, but I tested that the test case added works with a stage1.

r? @alexcrichton

**Comment by richo** [comment]
(make check passes locally, on osx)

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -844,12 +844,29 @@ impl Display for char {
...
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Pointer for *const T {
     fn fmt(&self, f: &mut Formatter) -> Result {
         f.flags |= 1 << (FlagV1::Alternate as u32);
+        f.flags |= 1 << (FlagV1::SignAwareZeroPad as u32);
+
+        let old_width = f.width;
+        if let None = f.width {
+            f.width = POINTER_PADDING;
+        }
```
Comment (alexcrichton) [inline src/libcore/fmt/mod.rs]: Could this change be left to another PR?

Comment (richo) [inline src/libcore/fmt/mod.rs]: Yup, np.

**Comment by richo** [comment]
Rebased, this is now only the fmt::Pointer implementation

**Comment by alexcrichton** [comment]
Thanks! Could you also squash the commits together? Other than that r=me

**Comment by richo** [comment]
Done, thanks!
